### PR TITLE
Users can edit their profile #83

### PR DESCRIPTION
**Pull request by shawki2**
The work is extended as follow:
If the user doesn't fill the password field, then the form can be submitted and the password will not be updated.
  If the user fills the password field, then the password will be updated.
  A message should be displayed near the Password label to inform the user that the password field should be filled only if the password needs to be updated.
If the user edits their email and chooses an email which already belongs to another existing user, an error should be displayed.

### CODE REVIEW

File: client/src/containers/App/App.js [hunk 1]
```diff
@@ -19,6 +19,7 @@ import EditLesson from "../Admin/Lessons/EditLesson/EditLesson";
 import ViewTopics from "../Admin/Lessons/Topics/ViewTopics";
 import AddTopics from "../Admin/Lessons/Topics/AddTopics/AddTopics";
 import EditTopics from "../Admin/Lessons/Topics/EditTopics/EditTopics";
+import EditUser from "../User/EditUser/editUser";
```
Comment (llh1) [inline client/src/containers/App/App.js]: Keep consistency in the naming: editUser -> EditUser

File: client/src/containers/User/EditUser/editUser.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      name: "",
+      password: "",
+      message: "",
+      messageAlert: ""
+    };
+  }
+
+  componentDidMount = async () => {
+    const user = await getUserProfile();
+    console.log(`Userdata= `, user);
```
Comment (llh1) [inline client/src/containers/User/EditUser/editUser.js]: You can remove console.log before you commit your code

File: client/src/containers/User/EditUser/editUser.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    this.setState({
+      user_id,
+      name,
+      email
+    });
+  };
+
+  handleOnchange = (input, e) => {
+    const value = e.target.value;
+    console.log("vlaue= ", value);
```
Comment (llh1) [inline client/src/containers/User/EditUser/editUser.js]: Remove console.log

File: server/migrations/20180811151440_topics.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ exports.up = async (knex, Promise) => {
   await knex.schema.createTable("users", table => {
     table.increments("user_id");
     table.string("name");
-    table.string("email");
+    table.string("email").unique();
```
Comment (llh1) [inline server/migrations/20180811151440_topics.js:49]: Good

File: server/helpers/db.js [hunk 5]
```diff
@@ -214,6 +218,10 @@ const checkUserByEmailExist = async email => {
   const response = await knex("users").where({ email });
   return response.length === 0 ? false : true;
 };
+const checkUserByIDExist = async user_id => {
+  const response = await knex("users").where({ user_id });
+  return response.length === 0 ? false : true;
+};
```
Comment (llh1) [inline server/helpers/db.js]: Try something like this:
`
const checkUserByIDExist = async (email, userId) => {
    const response = await knex("users").where('email', email).andWhereNot('user_id', userId);
    return response.length === 0 ? false : true;
};
`

File: server/routes/user.js [hunk 6]
```diff
@@ -18,6 +18,7 @@ router.get("/profile", async (req, res, next) => {
 router.put("/profile", async (req, res, next) => {
   const { user_id: userId } = req.user;
   const { name, email, password } = req.body;
+
```
Comment (llh1) [inline server/routes/user.js]: Use your function checkUserByIDExist here, and depending on the result, return an error to the client. Remember to use await with checkUserByIDExist.

File: server/migrations/20180811151440_topics.js (hunk 4 above)
Comment (shawki2) [reply server/migrations/20180811151440_topics.js:49]: Thanks for your comments.

File: server/helpers/db.js (hunk 5 above)
Comment (shawki2) [reply server/helpers/db.js]: Sorry, I have just seen this comment.

File: client/src/containers/User/EditUser/editUser.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    this.setState({
+      user_id,
+      name,
+      email
+    });
+  };
+
+  handleOnchange = (input, e) => {
+    const value = e.target.value;
+    this.setState({ [input]: value });
```
Comment (kabaros) [inline client/src/containers/User/EditUser/editUser.js:29]: nice use of es6 features

**Review by kabaros — APPROVED** [review]

**Pull request merged**